api/src/main/java/org/apache/iceberg/Table.java

```diff
@@ -51,6 +51,23 @@
    */
   Schema schema();
 
+  /**
+   * Return the {@link Schema schema} for this table at the time of the snapshot
+   * specified by the snapshotId.
+   *
+   * @return this table's schema
+   */
+  Schema schemaForSnapshot(long snapshotId);
+
+  /**
+   * Return the {@link Schema schema} for this table at the time of the most recent
+   * snapshot as of the specified timestampMillis.
+   * Note: This is not necessarily the schema at the time of the specified timestampMillis.
+   *
+   * @return this table's schema
+   */
+  Schema schemaForSnapshotAsOfTime(long timestampMillis);
```

Comment:
Instead of changing the table API in this commit, I think we only need to add a `schema` method to `Snapshot`. That will be used later when we have better tracking. Not adding too many methods to the public API in this commit will give us more flexibility later and make this simpler to get in.

Reply:
I can avoid adding the `schemaForSnaphot` methods to the `Table` API. I think only `BaseTable` needs to have the methods. However, the places where I need to call `schemaForSnapshot` have references to simply `Table`; I'd have to check if the `Table` is an instance of `BaseTable`.
I also suggested in the general comments that `Snapshot` could simply have a `schemaId` method in the future. And that would then get used in a revised implementation of `schemaForSnapshot` in `BaseTable`.

Reply:
Yeah, I agree with you. We have some flexibility here with how we look up the schema for a given snapshot. The main thing is that we don't want to expose any methods through the `Table` API right now. And since we don't have a schema id yet, we can't really add that method to `Snapshot`.

Eventually, I think we would want `Snapshot` to return its own schema rather than requiring a lookup, which is why I suggested that option. But we might be able to avoid the problem for now.

Like I said, I just don't think we want to add anything to `Table`.

Reply:
I have removed the methods from `Table`, adding them only to `BaseTable`. Thus the `Table` API is unchanged.